### DISCUSSION TITLE
How to replicate HopChar2 in mini.jump2d?

### DISCUSSION THREAD

**Discussion by justEstif** [nvim-mini/mini.nvim#205] [Q&A] [answered]
Hello,

I have been using the builtin jump2d opts, but I have found writing my own spotter challenging.
If you don't mind, could you assist me in writing it? 

- I had tried to copy parts of the code for builtin opts, but I don't know how to limit the user input to two characters.

I have also found the builtin jump by query option to be great, but it is case-sensitive. Is there a way to make it case-insensitive?

Thanks for the awesome plugin!

**Comment by echasnovski** [comment]
Hi!

> I had tried to copy parts of the code for builtin opts, but I don't know how to limit the user input to two characters.

My first suggestion is to look in direction of [`MiniJump2d.builtin_opts.single_character`](https://github.com/echasnovski/mini.nvim/blob/305debd01d486f94a9f6ecfec90321ea98dc5dd9/lua/mini/jump2d.lua#L518) (more specifically, [this helper](https://github.com/echasnovski/mini.nvim/blob/305debd01d486f94a9f6ecfec90321ea98dc5dd9/lua/mini/jump2d.lua#L493)) and ask user twice for an input character. Use `vim.fn.getcharstr` instead of `input_fun` (or copy code for `H.getcharstr()`.

> I have also found the builtin jump by query option to be great, but it is case-sensitive. Is there a way to make it case-insensitive?

It should be possible, although a bit hacky. Again, take a look at [this helper]([this helper](https://github.com/echasnovski/mini.nvim/blob/305debd01d486f94a9f6ecfec90321ea98dc5dd9/lua/mini/jump2d.lua#L493)). Use `vim.fn.input` instead of `input_fun` and modify its output to model case insensitivity with Lua pattern. Like if output is `ab` then "case-insensitive" pattern should be `[aA][bB]`. Use Lua's `string.upper()` and `string.lower()` to achieve it programmatically.

I am not sure when I'd be able to write a full perfectly working example of `HopChar2` for you, sorry.

**Reply by justEstif** [reply]
Thanks for the input. I will try do do what you suggested and post the result here.

**Comment by echasnovski** [comment, marked as answer]
Ok, I've got time sooner than I thought.

Try this snippet:

```lua
local safe_getcharstr = function(msg)
  vim.cmd('echon ' .. vim.inspect(msg))

  -- Allow `<C-c>` to end input
  local ok, res = pcall(vim.fn.getcharstr)

  -- Clean command line
  vim.cmd([[echo '' | redraw]])

  -- Treat `<Esc>` or `<CR>` as cancel
  if not ok or (res == '\27' or res == '\r') then return nil end

  return res
end

local make_ignorecase_pattern = function(word)
  local parts = {}
  for i = 1, word:len() do
    local char = word:sub(i, i)

    if char:find('^%a$') then
      -- Convert letter to a match both lower and upper case
      char = '[' .. char:lower() .. char:upper() .. ']'
    else
      -- Escape non-letter characters
      char = vim.pesc(char)
    end

    table.insert(parts, char)
  end

  return table.concat(parts)
end

local dummy_spotter = function() return {} end

require('mini.jump2d').setup({
  spotter = dummy_spotter,
  allowed_lines = { blank = false, fold = false },
  hooks = {
    before_start = function()
      local first = safe_getcharstr("(mini.jump2d) Enter first character: ")
      if first == nil then
        MiniJump2d.config.spotter = dummy_spotter
        return
      end

      local second = safe_getcharstr("(mini.jump2d) Enter second character: ")
      if second == nil then
        MiniJump2d.config.spotter = dummy_spotter
        return
      end

      local pattern = make_ignorecase_pattern(first .. second)
      MiniJump2d.config.spotter = MiniJump2d.gen_pattern_spotter(pattern)
    end,
  },
})
```

**Reply by justEstif** [reply]
That worked, thanks a lot.
